Rewrite v3 api

UpcraftLP opened the pull request:
Fully rewrite how data is gathered and added to routes, while maintaining full compatibility with the existing v3 api (no format changes or removals until api v4).

- additionally provide hashes for maven artifacts EXCEPT intermediary, this is a backwards compatible change; hashes are pulled from gradle module metadata, then falling back to maven hash files, and (since technically both of these are optional) finally falling back to downloading the actual file and calculating the hash manually
- new manifest format that is more easily parseable (json) should that ever be necessary; this however means that on the initial run ALL files will be flagged as outdated and reuploaded
- fixed timestamp format for launcher profiles to be in line with mojang's date formatting (`+00:00` instead of `+0000`); this is not a breaking change since the iso standard explicitly defines both versions as valid, so any ISO8601-compliant parser will handle this fine.
- due to the above, and installer profile json release dates relying on cache hits. ~~this will cause all of these to have a release time of whenever the v2 manifest gets generated.~~ For now we simply copy the release times of the respective MC version. This should be fine but ideally we would track loader release times separately and then calculate the release time as `max(loaderReleaseTime, minecraftVersionReleaseTime)`. Probably better left for v4 tho.
- support for not using cloudflare cache, plus abstractions to add in other cache mechanisms if needed
- storage support for local directory (mostly for debug, this will append file extensions to routes because due to the v3 route architecture we would otherwise have folders and files with the same paths, which is illegal on most file systems)
- storage support for s3
- storage support for backblaze b2
- full CLI interface for the updater tool to switch between backends, caches, and provide all the necessary credentials and validation thereof
- overhauled GitHub actions scripts to do a test run and export the outputs as job artifacts

supersedes #21

UpcraftLP committed:
move to new package

UpcraftLP committed:
use java 21 and update GH workflows

UpcraftLP committed:
use gradle version catalog

UpcraftLP committed:
add java application plugin

UpcraftLP committed:
use hex instead of base64 for index hashes and handle duplicate files

UpcraftLP committed:
*record scratch*

*freeze frame* Yep, that's me. You're probably wondering how I got here.

UpcraftLP committed:
add hash to file upload objects

UpcraftLP committed:
full rewrite

UpcraftLP committed:
cleanup

UpcraftLP committed:
add custom hash provider for future use

UpcraftLP committed:
run build action on all branches

UpcraftLP committed:
export generated image with more tags

UpcraftLP committed:
fix docker build args

UpcraftLP committed:
run gradle tests on build

UpcraftLP committed:
output JUnit reports to workflow log

UpcraftLP committed:
*french accent* This... is a record

(*dear god*) there's more.. (*no*) it contains an automatic equals and hashcode implementation

UpcraftLP committed:
use regular sha tag instead of prefixed one

UpcraftLP committed:
run build on pull requests, too

UpcraftLP committed:
remove entrypoint hack

UpcraftLP committed:
use new message digest for every hash

MessageDigest is not thread-safe

UpcraftLP committed:
fix dockerfile app name

UpcraftLP committed:
download existing hashes from maven

UpcraftLP committed:
add file sizes and hashes (closes #5)

UpcraftLP committed:
less verbose logging for gradle module files

UpcraftLP committed:
increase maven cache size

UpcraftLP committed:
update OpenAPI spec

(supersedes #21)

UpcraftLP committed:
generate debug artifacts

UpcraftLP committed:
fix using wrong variable

UpcraftLP committed:
merge shadowJar build into regular build step

UpcraftLP committed:
list files

UpcraftLP committed:
test?

UpcraftLP committed:
fix jar name

UpcraftLP committed:
better compression for debug files

UpcraftLP committed:
exopand minecraft meta

parity with installer

UpcraftLP committed:
filter intermediary by hashed availability

UpcraftLP committed:
clarify supported versions

UpcraftLP committed:
update logger config

eristhea reviewed (approved):
i like the way you organised this, it's much better than the old code! i also appreciate all the notes for v4 and the new logging :)

UpcraftLP committed:
pass mc release times for launcher profiles

UpcraftLP committed:
hide experimental snapshots

maintain parity with v3 meta and only surface experimental snapshot versions in intermediary version endpoint

UpcraftLP committed:
temp fix to make file based storage not crash
